pkg/engine/renderer/dependencies.go

```diff
@@ -0,0 +1,245 @@
+	c.cache[typeConfigMap] = map[string]hashBytes{}
+	return c
+}
+
+var (
+	// The types of dependencies we support
+	typeSecret    = reflect.TypeOf(corev1.Secret{})
+	typeConfigMap = reflect.TypeOf(corev1.ConfigMap{})
+
+	// We need a list of types to ensure correct order of hash calculation
+	dependencyTypes = []reflect.Type{
+		typeSecret,
+		typeConfigMap,
+	}
+)
+
+type resourceDependencies map[reflect.Type][]string
+
+func newDependencies() resourceDependencies {
+	deps := map[reflect.Type][]string{}
+
+	for _, t := range dependencyTypes {
+		deps[t] = []string{}
+	}
+
+	return deps
+}
+
+// Adds all dependencies from a pod template spec
+func (rd resourceDependencies) addFromPodTemplateSpec(SpecTemplate corev1.PodTemplateSpec) {
+	for _, s := range SpecTemplate.Spec.ImagePullSecrets {
+		rd[typeSecret] = append(rd[typeSecret], s.Name)
+	}
+	for _, v := range SpecTemplate.Spec.Volumes {
+		if v.ConfigMap != nil {
+			rd[typeConfigMap] = append(rd[typeConfigMap], v.ConfigMap.Name)
+		}
+		if v.Secret != nil {
+			rd[typeSecret] = append(rd[typeSecret], v.Secret.SecretName)
+		}
+	}
+}
+
+// Adds a hash calculated from the dependencies to embedded pod template specs
+func (de *dependencyCalculator) calculateAndSetHash(obj metav1.Object, deps resourceDependencies) error {
+
+	depHash := md5.New() //nolint:gosec
+	for _, depType := range dependencyTypes {
+		for _, name := range deps[depType] {
+			hash, err := de.getHashForDependency(name, depType)
+			if err != nil {
+				return fmt.Errorf("error calculating hash for %s of type %s: %v", name, depType, err)
+			}
+			_, _ = depHash.Write(hash[:])
+		}
+	}
+
+	hashStr := fmt.Sprintf("%x", depHash.Sum([]byte{}))
+
+	return setTemplateHash(obj, hashStr)
+}
+
+func (de *dependencyCalculator) getHashForDependency(name string, t reflect.Type) (hashBytes, error) {
+	cache := de.cache[t]
+
+	if hash, ok := cache[name]; ok {
+		return hash, nil
+	}
+
+	dep, err := de.resourceDependency(name, t)
+	if err != nil {
+		return hashBytes{}, fmt.Errorf("failed to get dependeny %s/%s: %v", de.namespace, name, err)
+	}
+	if _, ok := dep.GetAnnotations()[kudo.SkipHashCalculationAnnotation]; ok {
+		cache[name] = hashBytes{}
+	} else {
+		yamlStr, err := calculateResourceHash(dep)
+		if err != nil {
+			return hashBytes{}, fmt.Errorf("failed to serialize dependeny %s/%s: %v", de.namespace, name, err)
+		}
+		cache[name] = md5.Sum([]byte(yamlStr)) //nolint:gosec
+	}
+
+	return cache[name], nil
+}
+
+// Calculates a stable hash from a resource
+func calculateResourceHash(obj metav1.Object) (string, error) {
+	// Namespace is ignored mostly to allow easier integration tests (which use random namespaces)
+	ns := obj.GetNamespace()
+
+	// OwnerReferences need to be skipped as they contain a changing UID
+	or := obj.GetOwnerReferences()
+
+	// Annotations are notorious for containing quickly changing strings: plan/phase/task names, uids, dates, etc.
+	ann := obj.GetAnnotations()
```

Comment:
Good question. I think ignoring all annotations here is ok, at least for now. At the moment, dependencies are ConfigMaps and Secrets, and we only care if the actual content changes. 

This might be different in the future if/when we add different kinds of dependencies. I don't really see a use case for that at the moment, but i guess it's possible.